File: 0401-0500/0426-Convert Binary Search Tree to Sorted Doubly Linked List/0426-Convert Binary Search Tree to Sorted Doubly Linked List.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;

    Node() {}

    Node(int _val) {
        val = _val;
        left = NULL;
        right = NULL;
    }

    Node(int _val, Node* _left, Node* _right) {
        val = _val;
        left = _left;
        right = _right;
    }
};
*/

class Solution {
public:
    Node* treeToDoublyList(Node* root) {
        if (root == nullptr) {
            return root;
        }
        stack<Node*> St;
        Node *curr = root, *prev = nullptr, *head = nullptr, *tail = nullptr;
        while (!St.empty() || curr) {
            if (curr) {
                St.push(curr);
                curr = curr->left;
            }
            else {
                curr = St.top();
                St.pop();
                if (prev == nullptr) {
                    head = curr;
                }
                else {
                    curr->left = prev;
                    prev->right = curr;
                }
                tail = curr;
                prev = curr;
                curr = curr->right;
            }
        }
        
        head->left = tail;
        tail->right = head;
        return head;
    }
};

File: 0401-0500/0426-Convert Binary Search Tree to Sorted Doubly Linked List/0426-Convert Binary Search Tree to Sorted Doubly Linked List.py
# Definition for a Node.
class Node:
    def __init__(self, val, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def treeToDoublyList(self, root: 'Node') -> 'Node':
        if root is None:
            return None
        St = []
        head = tail = prev = None
        curr = root
        while St or curr:
            if curr:
                St.append(curr)
                curr = curr.left
            else:
                curr = St.pop()
                if prev is None:
                    head = curr
                else:
                    curr.left = prev
                    prev.right = curr
                
                tail = curr
                prev = curr
                curr = curr.right
        
        head.left = tail
        tail.right = head
        return head
